### PR TITLE
Add realpath function

### PR DESCRIPTION
**Pull request by fourier**
1. Redefined stat structure for OSX
2. Added realpath function

### CODE REVIEW

**Comment by fourier** [comment]
Looks like the repo has its compilation problems with clisp and ccl even before my pull request.

**Comment by fourier** [comment]
Is there anyone who can take a look at this?

File: posix/basic-unix.lisp [hunk 1]
```diff
@@ -223,3 +223,18 @@
...
+
+(defun realpath (path)
+  "Returns the canonicalized absolute pathname"
+  (with-foreign-pointer (resolved-name path-max)
+    (let ((result (%realpath path resolved-name)))
+      (when (not (null-pointer-p result))
+        (multiple-value-bind (string length)
+            (foreign-string-to-lisp resolved-name)
+          (declare (ignore length))
+          string)))))
```
Comment (luismbo) [inline posix/basic-unix.lisp]: `foreign-string-to-lisp` returns `nil` when passed a null pointer so you could simplify this to

```lisp
(with-foreign-pointer (resolved-name path-max)
  (foreign-string-to-lisp (%realpath path resolved-name)))
```

Or you could specify `%realpath` to have a `:string` return type and remove the call to `foreign-string-to-lisp`, I think.

File: posix/basic-unixint.lisp [hunk 2]
```diff
@@ -312,3 +318,24 @@
...
+  (mtimespec     "st_mtimespec"     :type timespec)        ;; time of last data modification 
+  (ctimespec     "st_ctimespec"     :type timespec)        ;; time of last status change 
+  (birthtimespec "st_birthtimespec" :type timespec)        ;; time of file creation(birth) 
+  (size          "st_size"          :type off)             ;; file size, in bytes 
+  (blocks        "st_blocks"        :type blkcnt)          ;; blocks allocated for file 
+  (blksize       "st_blksize"       :type blksize)         ;; optimal blocksize for I/O 
+  (flags         "st_flags"         :type :uint32)         ;; user defined flags for file 
+  (gen           "st_gen"           :type :uint32)         ;; file generation number 
+  (lspare        "st_lspare"        :type :int32)          ;; RESERVED: DO NOT USE! 
+  (qspare        "st_qspare"        :type :int64 :count 2));; RESERVED: DO NOT USE!
```
Comment (luismbo) [inline posix/basic-unixint.lisp]: Why define reserved fields?

**Comment by luismbo** [comment]
Also, can you use `defsyscall` so that when `realpath()` fails we look at errno and signal a condition?

**Comment by fourier** [comment]
Hi,
Thanks for comments. I will split into 2 pull requests. In this request I'll just add realpath function. Indeed the usage of :string type simplifies things. As for defsyscall I already do it for the original function ```%realpath```, the ```realpath``` function is just a wrapper around this syscall. It indeed already signals errors, I've already tested it.

**Comment by luismbo** [comment]
Oh, I missed that you were already using `defsyscall` somehow. I think I was confused by your explicit checking for nil since `defsyscall` takes care of that, IIRC.

**Comment by luismbo** [comment]
#24

**Comment by fourier** [comment]
I've already split to 2 pull requests and updated both according to comments, please have a look!

**Comment by fourier** [comment]
Updated commit text.

**Comment by sionescu** [comment]
Merged, thanks.

**Pull request merged**